content/posts/expose-serial-port.md

```diff
@@ -0,0 +1,190 @@
+ 
+   // This is the number of virtual disks (block devices) that should be
+   // configured by the VmManager. Related to the description above regarding
+```
+
+And finally let's actually forward the new parameter to `crosvm`:
+```
+diff --git a/host/libs/vm_manager/crosvm_manager.cpp b/host/libs/vm_manager/crosvm_manager.cpp
+index 100b71e56..3f568ce28 100644
+--- a/host/libs/vm_manager/crosvm_manager.cpp
++++ b/host/libs/vm_manager/crosvm_manager.cpp
+@@ -449,6 +449,20 @@ Result<std::vector<MonitorCommand>> CrosvmManager::StartCommands(
+   for (auto i = 0; i < VmManager::kMaxDisks - disk_num; i++) {
+     crosvm_cmd.AddHvcSink();
+   }
++
++  if (!config.attach_serial_device().empty()) {
++    crosvm_cmd.AddHvcReadWrite(config.attach_serial_device(),
++                               config.attach_serial_device());
++    LOG(INFO) << "SERIAL_PORT: attaching serial port to crosvm: "
++              << config.attach_serial_device()
++              << ", most likely device name inside the guest machine: /dev/hvc"
++              << crosvm_cmd.HvcNum() - 1;
++  } else {
++    crosvm_cmd.AddHvcSink();
++  }
++
+   CF_EXPECT(crosvm_cmd.HvcNum() + disk_num ==
+                 VmManager::kMaxDisks + VmManager::kDefaultNumHvcs,
+             "HVC count (" << crosvm_cmd.HvcNum() << ") + disk count ("
+```
+
+... but not to QEMU, which is mostly used as AVD for normal (phone) Android apps (it might actually work, but we couldn't test because of unrelated problems):
+```
+diff --git a/host/libs/vm_manager/qemu_manager.cpp b/host/libs/vm_manager/qemu_manager.cpp
+index a2744a35d..e6b23d2d5 100644
+--- a/host/libs/vm_manager/qemu_manager.cpp
++++ b/host/libs/vm_manager/qemu_manager.cpp
+@@ -556,6 +556,8 @@ Result<std::vector<MonitorCommand>> QemuManager::StartCommands(
+     add_hvc_sink();
+   }
+ 
++  add_hvc_sink();  // attach_serial_device, not supported on QEMU
++
+   CF_EXPECT(
+       hvc_num + disk_num == VmManager::kMaxDisks + VmManager::kDefaultNumHvcs,
+       "HVC count (" << hvc_num << ") + disk count (" << disk_num << ") "
+```
+
+And the last thing we need to do is setting up proper access rights:
+```
+diff --git a/shared/config/ueventd.rc b/shared/config/ueventd.rc
+index c604e3d23..349f8fdcb 100644
+--- a/shared/config/ueventd.rc
++++ b/shared/config/ueventd.rc
+@@ -44,5 +44,13 @@
+ # keymint / Rust
+ /dev/hvc11 0666 system system
+ 
++# Serial Port integration for the -attach_serial_device
++# param of the launch_cvd command.
++# Please update this name "/dev/hvcN" to "/dev/hvc{N+M}"
++# when kDefaultNumHvcs in 'host/libs/vm_manager/vm_manager.h' is updated.
++# The 'M' value here would be the delta between the new and the old values
++# of kDefaultNumHvcs.
++/dev/hvc12 0666 system system
++
+ # Factory Reset Protection
+ /dev/block/by-name/frp 0660 system system
+```
+
+# Conclusion
+And that was basically it - now you can use the exposed port (`/dev/hvc12` in the case of the diff above) from inside of Android.
```

Comment:
a full launch_cvd command as an example would be nice

Reply:
Also you can upload a full A14 patch in our github repo, so people can easily apply it